File: lbcli-rs/src/debug.rs
use structopt::StructOpt;

use crate::util::file_from_path_or_id;
use crate::CliError;

#[derive(StructOpt)]
pub enum DebugCmd {
    /// Prints metadata of a lockbook file via path or ID
    Finfo(FinfoArgs),
    /// Helps find invalid states within lockbook
    Validate,
}

#[derive(StructOpt)]
pub struct FinfoArgs {
    /// Lockbook file path or ID
    target: String,
    /// Print out the raw Rust `Debug` text of the lockbook file
    #[structopt(long)]
    raw: bool,
}

pub fn debug(core: &lb::Core, cmd: DebugCmd) -> Result<(), CliError> {
    match cmd {
        DebugCmd::Finfo(args) => finfo(core, args),
        DebugCmd::Validate => validate(core),
    }
}

fn finfo(core: &lb::Core, args: FinfoArgs) -> Result<(), CliError> {
    let my_name = core.get_account()?.username;
    match file_from_path_or_id(core, &args.target) {
        Ok(f) => print_file(&f, args.raw, &my_name),
        Err(err) => {
            let ids = core
                .list_metadatas()
                .map_err(|err| CliError(format!("listing metadatas: {:?}", err)))?
                .into_iter()
                .filter_map(|f| {
                    if f.id.to_string().starts_with(&args.target) {
                        Some(f.id)
                    } else {
                        None
                    }
                })
                .collect::<Vec<lb::Uuid>>();
            match ids.len() {
                0 => return Err(err),
                1 => {
                    let id = ids[0];
                    let f = core.get_file_by_id(id).map_err(|err| (err, id))?;
                    print_file(&f, args.raw, &my_name);
                }
                _ => {
                    println!("id prefix matched the following files:\n");
                    for id in ids {
                        let msg = match core.get_path_by_id(id) {
                            Ok(p) => p,
                            Err(err) => format!("error getting path: {:?}", err),
                        };
                        println!(" {}  {}", id, msg);
                    }
                }
            }
        }
    }
    Ok(())
}

fn print_file(f: &lb::File, raw: bool, my_name: &str) {
    if raw {
        println!("{:#?}", f);
        return;
    }
    // Build the text that will contain share info.
    let mut shares = String::new();
    for sh in &f.shares {
        let shared_by = if sh.shared_by == my_name {
            String::from("me")
        } else {
            format!("@{}", sh.shared_by)
        };
        let shared_with = if sh.shared_with == my_name {
            String::from("me")
        } else {
            format!("@{}", sh.shared_with)
        };
        shares += &format!("\n    {} -> {} ({:?})", shared_by, shared_with, sh.mode);
    }
    let data = &[
        ("name", f.name.to_string()),
        ("id", f.id.to_string()),
        ("parent", f.parent.to_string()),
        ("type", f.file_type.to_string().to_lowercase()),
        ("lastmod", f.last_modified.to_string()),
        ("lastmod_by", f.last_modified_by.to_string()),
        (&format!("shares ({})", f.shares.len()), shares),
    ];
    // Determine widest key name.
    let mut w_name = 0;
    for (k, _) in data {
        let n = k.len();
        if n > w_name {
            w_name = n;
        }
    }
    for (k, v) in data {
        println!("  {:<w_name$} : {}", k, v, w_name = w_name);
    }
}

fn validate(core: &lb::Core) -> Result<(), CliError> {
    let warnings = core
        .validate()
        .map_err(|err| CliError(format!("validating: {:?}", err)))?;
    if warnings.is_empty() {
        return Ok(());
    }
    for w in &warnings {
        eprintln!("{:#?}", w);
    }
    Err(CliError(format!("{} warnings found", warnings.len())))
}

File: lbcli/debug.go
package main

import (
	"fmt"
	"os"
	"strings"

	"github.com/steverusso/lockbook-x/go-lockbook"
)

// Investigative commands mainly intended for devs.
type debugCmd struct {
	finfo    *debugFinfoCmd
	validate *debugValidateCmd
	whoami   *debugWhoamiCmd
}

func (d *debugCmd) run(core lockbook.Core) error {
	switch {
	case d.finfo != nil:
		return d.finfo.run(core)
	case d.validate != nil:
		return d.validate.run(core)
	case d.whoami != nil:
		return d.whoami.run(core)
	default:
		return nil
	}
}

// View info about a target file.
type debugFinfoCmd struct {
	// The target can be a file path, UUID, or UUID prefix.
	//
	// clap:arg_required
	target string
}

func (c *debugFinfoCmd) run(core lockbook.Core) error {
	id, err := idFromSomething(core, c.target)
	if err != nil {
		return fmt.Errorf("trying to get id from %q: %w", c.target, err)
	}
	f, err := core.FileByID(id)
	if err != nil {
		return fmt.Errorf("getting file %q: %w", id, err)
	}
	acct, err := core.GetAccount()
	if err != nil {
		return fmt.Errorf("getting account: %w", err)
	}
	printFile(f, acct.Username)
	return nil
}

func printFile(f lockbook.File, myName string) {
	// Build the text that will contain share info.
	shares := ""
	for _, sh := range f.Shares {
		sharedBy := "@" + sh.SharedBy
		if sh.SharedBy == myName {
			sharedBy = "me"
		}
		sharedWith := "@" + sh.SharedWith
		if sh.SharedWith == myName {
			sharedWith = "me"
		}
		mode := strings.ToLower(sh.Mode.String())
		shares += fmt.Sprintf("\n    %s -> %s (%s)", sharedBy, sharedWith, mode)
	}
	data := [][2]string{
		{"name", f.Name},
		{"id", f.ID.String()},
		{"parent", f.Parent.String()},
		{"type", strings.ToLower(lockbook.FileTypeString(f.Type))},
		{"lastmod", fmt.Sprintf("%v", f.Lastmod)},
		{"lastmod_by", f.LastmodBy},
		{fmt.Sprintf("shares (%d)", len(f.Shares)), shares},
	}
	// Determine widest key name.
	nameWidth := 0
	for i := range data {
		n := len(data[i][0])
		if n > nameWidth {
			nameWidth = n
		}
	}
	for i := range data {
		fmt.Printf("  %-*s : %s\n", nameWidth, data[i][0], data[i][1])
	}
}

// Find invalid states within your lockbook.
type debugValidateCmd struct{}

func (debugValidateCmd) run(core lockbook.Core) error {
	warnings, err := core.Validate()
	if err != nil {
		return fmt.Errorf("running validate: %w", err)
	}
	count := len(warnings)
	if count == 0 {
		return nil
	}
	suffix := ""
	if count > 1 {
		suffix = "s"
	}
	fmt.Fprintf(os.Stderr, "\033[1;33m%d warning%s found:\033[0m\n", count, suffix)
	for _, w := range warnings {
		fmt.Fprintf(os.Stderr, "  %s\n", w)
	}
	return nil
}

// Print user information for this lockbook.
type debugWhoamiCmd struct{}

func (debugWhoamiCmd) run(core lockbook.Core) error {
	acct, err := core.GetAccount()
	if err != nil {
		return fmt.Errorf("getting account: %w", err)
	}
	fmt.Printf("data-dir: %s\n", core.WriteablePath())
	fmt.Printf("username: %s\n", acct.Username)
	fmt.Printf("server:   %s\n", acct.APIURL)
	return nil
}
